middle_end/flambda/simplify/simplify_extcall.ml

```diff
@@ -0,0 +1,159 @@
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+[@@@ocaml.warning "+a-30-40-41-42"]
+
+(* The matches on type proofs are annoying to write exhaustively. *)
+[@@@ocaml.warning "-4"]
+
+open! Simplify_import
+
+type t =
+  | Unchanged
+  | Poly_compare_specialized of DA.t * Expr.t
+
+(* Helpers *)
+(* ******* *)
+
+let fun_symbol simple =
+  let fail simple =
+    Misc.fatal_errorf
+      "Expected a function symbol, instead of@ %a" Simple.print simple
+  in
+  Simple.pattern_match simple
+    ~name:(fun name ->
+      Name.pattern_match name
+        ~var:(fun _ -> fail simple)
+        ~symbol:(fun sym -> sym))
+    ~const:(fun _ -> fail simple)
+
+let simple_of_var v =
+  Simple.name (Name.var v)
+
+let apply_cont cont v ~dbg =
+  let args = [ Simple.name (Name.var v) ] in
+  let apply_cont = Apply_cont.create cont ~args ~dbg in
+  let free_names = Apply_cont.free_names apply_cont in
+  let expr = Expr.create_apply_cont apply_cont in
+  free_names, expr
+
+let let_prim ~dbg v prim (free_names, body) =
+  let v' = Var_in_binding_pos.create v Name_mode.normal in
+  let bindable = Bindable_let_bound.singleton v' in
+  let named = Named.create_prim prim dbg in
+  let free_names_of_body = Or_unknown.Known free_names in
+  let let_expr = Let.create bindable named ~body ~free_names_of_body in
+  let free_names = Name_occurrences.remove_var free_names v in
+  let expr = Expr.create_let let_expr in
+  free_names, expr
+
+
+(* Exported simplification function *)
+(* ******************************** *)
+
+let simplify_comparison ~dbg ~dacc ~cont
+      ~int_prim ~float_prim a b a_ty b_ty =
+  let tenv = DA.typing_env dacc in
+  match T.prove_is_a_tagged_immediate tenv a_ty,
+        T.prove_is_a_tagged_immediate tenv b_ty with
+  | Proved (), Proved () ->
+    let v_comp = Variable.create "comp" in
+    let tagged = Variable.create "tagged" in
+    let _, res =
+      let_prim ~dbg v_comp (P.Binary (int_prim, a, b)) @@
+      let_prim ~dbg tagged
+        (P.Unary (Box_number Untagged_immediate, simple_of_var v_comp)) @@
+      apply_cont ~dbg cont tagged
+    in
+    Poly_compare_specialized (dacc, res)
+  | _ ->
+    begin match T.prove_is_a_boxed_float tenv a_ty,
+                T.prove_is_a_boxed_float tenv b_ty with
+    | Proved (), Proved () ->
+      let a_naked = Variable.create "naked_float" in
+      let b_naked = Variable.create "naked_float" in
+      let v_comp = Variable.create "comp" in
+      let tagged = Variable.create "tagged" in
+      let _, res =
```

Comment:
_ => _free_names otherwise it's really not obvious what it is without looking at more code

Reply:
done in e9f3989